### PR TITLE
Add Utils for federated learning scenarios

### PR DESCRIPTION
**Pull request by AdamLouly**
**Description**: utils for federated learning.

**Motivation and Context**
- This PR includes utils that will be used on federated learning scenarios.
- Exposing python bindings to some utils, and added a util to calculate the difference between two buffers.

### CODE REVIEW

**Comment by justinchuby** [comment]
I would use a more descriptive PR title

File: orttraining/orttraining/python/training/api/on_device_utils.py [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+# Licensed under the MIT License.
+# checkpoint_utils.py
+
+import numpy as np
+
+from onnxruntime.capi import _pybind_state as C
+from onnxruntime.capi._pybind_state import get_parameters_difference as _internal_get_parameters_difference
+
+
+def get_parameters_difference(output_params, old_output_params):
```
Comment (justinchuby) [inline orttraining/orttraining/python/training/api/on_device_utils.py:11]: Prefer type annotating any new functions

File: orttraining/orttraining/python/training/api/module.py [hunk 2]
```diff
@@ -81,3 +83,33 @@ def save_checkpoint(self, ckpt_uri):
...
+            ],
+            np.float32,
+            "cpu",
+            0,
+        )._ortvalue
+        self._model.copy_parameters_to_buffer(parameters)
+
+        return parameters
+
+    def get_parameters_size(self, trainable_only: bool = False) -> int:
```
Comment (baijumeswani) [inline orttraining/orttraining/python/training/api/module.py:105]: Do we need to expose this function here if `get_contiguous_parameters` is already allocating the buffer for the user? When would the user need this?

**Review by askhade — APPROVED** [review]
LGTM

File: orttraining/orttraining/python/training/api/module.py (hunk 2 above)
Comment (AdamLouly) [reply orttraining/orttraining/python/training/api/module.py:105]: This one is needed in the Federated learning scenarios where getting size is needed

Comment (baijumeswani) [reply orttraining/orttraining/python/training/api/module.py:105]: When would the user need to get the size? I think it is fine, just wanted to understand the context behind this.

File: orttraining/orttraining/python/training/api/module.py [hunk 3]
```diff
@@ -81,3 +83,33 @@ def save_checkpoint(self, ckpt_uri):
...
+        """
+        size = self._model.get_parameters_size(trainable_only)
+        parameters = OrtValue.ortvalue_from_shape_and_type(
+            [
+                size,
+            ],
+            np.float32,
+            "cpu",
+            0,
+        )._ortvalue
```
Comment (baijumeswani) [inline orttraining/orttraining/python/training/api/module.py:100]:
```suggestion
        parameters = OrtValue.ortvalue_from_shape_and_type(
            [
                self.get_parameters_size(trainable_only),
            ],
            np.float32,
            "cpu",
            0,
        )._ortvalue
```

**Review by baijumeswani — APPROVED** [review]

**Comment by AdamLouly** [comment]
Thanks for the review

**Pull request merged**